Sources/PackageGraph/PinsStore.swift

```diff
@@ -84,93 +72,185 @@ public final class PinsStore {
     /// - Parameters:
     ///   - packageRef: The package reference to pin.
     ///   - state: The state to pin at.
-    public func pin(
-        packageRef: PackageReference,
-        state: CheckoutState
-    ) {
-        self.pinsMap[packageRef.identity] = Pin(
+    public func pin(packageRef: PackageReference, state: CheckoutState) {
+        self.add(.init(
             packageRef: packageRef,
             state: state
-        )
+        ))
     }
 
     /// Add a pin.
     ///
     /// This will replace any previous pin with same package name.
     public func add(_ pin: Pin) {
-        self.pinsMap[pin.packageRef.identity] = pin
+        self._pins[pin.packageRef.identity] = pin
     }
 
     /// Unpin all of the currently pinned dependencies.
     ///
     /// This method does not automatically write to state file.
     public func unpinAll() {
         // Reset the pins map.
-        self.pinsMap = [:]
+        self._pins.clear()
     }
 
     public func saveState() throws {
-        if self.pinsMap.isEmpty {
+        try self.storage.save(pins: self._pins.get(), mirrors: self.mirrors, removeIfEmpty: true)
+    }
+}
+
+// MARK: - Serialization
+
+fileprivate struct PinsStorage {
+    private let path: AbsolutePath
+    private let lockFilePath: AbsolutePath
+    private let fileSystem: FileSystem
+    private let encoder = JSONEncoder.makeWithDefaults()
+    private let decoder = JSONDecoder.makeWithDefaults()
```

Comment:
Is there any particular advantage to keeping these around?  (e.g. performance)  Is there any concern about unintentional leaking of state from one encode/decode to the next?

Reply:
memory and performance. I dont think they hold any state.